src/kOS/Suffixed/Part/EngineValue.cs

```diff
@@ -213,6 +246,215 @@ public ScalarValue GetPossibleThrustAtAtm(ScalarValue atmPressure)
+                }
+            }
+
+            return resources;
+        }
+
+        public StringValue GetEngineConfig()
+        {
+            if (EngineConfig != null)
+            {
+                var configField = EngineConfig.GetType().GetField("configuration");
+                if (configField != null)
+                    return (string)configField.GetValue(EngineConfig);
+            }
+
+            return Part.partInfo.title;
+        }
+
+        public ScalarValue GetFuelFlow()
+        {
+            // fuelFlowGui is overidden by RealFuels to display massflow, so work around it.
+            if (HasRealFuels)
+            {
+                double fuelFlow = 0;
+
+                foreach (ModuleEngines e in FilteredEngineList)
+                {
+                    // currentAmount does not update when there is no fuel flow, so only update when fuel is flowing.
+                    if (e.fuelFlowGui > 0.0f)
+                    {
+                        foreach (Propellant p in e.propellants)
+                        {
+                            fuelFlow += p.currentAmount / Time.fixedDeltaTime;
+                        }
+                    }
+                }
+
+                return fuelFlow;
+            }
+            else
+            {
+                // Stock just shows flow.
+                return FilteredEngineList.Sum(e => e.fuelFlowGui);
+            }
+        }
+
+        // Unfortunately mixtureDensity is unreliable so there isn't a quick way to get this.
+        public ScalarValue GetMaxFuelFlow()
+        {
+            double maxFuelFlow = 0;
+
+            foreach (ModuleEngines e in FilteredEngineList)
+            {
+                foreach (Propellant p in e.propellants)
+                {
+                    maxFuelFlow += e.getMaxFuelFlow(p);
+                }
+            }
+
+            return maxFuelFlow;
+        }
+
+        public ScalarValue GetMassFlow()
+        {
+            double massFlow = 0;
+
+            foreach (ModuleEngines e in FilteredEngineList)
+            {
+                // currentAmount does not update when there is no fuel flow, so only update when fuel is flowing.
+                if (e.fuelFlowGui > 0.0f)
+                {
+                    foreach (Propellant p in e.propellants)
+                    {
+                        massFlow += p.currentAmount * p.resourceDef.density / Time.fixedDeltaTime;
+                    }
+                }
+            }
+
+            return massFlow;
+        }
+
+        public BooleanValue GetUllage()
+        {
+            bool ullage = false;
+
+            if (HasRealFuels)
+            {
+                foreach (ModuleEngines e in FilteredEngineList)
+                {
+                    var ullageField = e.GetType().GetField("ullage");
+                    if (ullageField != null)
+                    {
+                        ullage |= (bool)ullageField.GetValue(e);
```

Comment:
Is there a reason this uses bitwise OR rather than Boolean OR?  I guess because there's no `||=` operator in C#, but if the goal is to short-circuit once `ullage` is true, wouldn't it be better to just completely abort the `foreach` loop once it's true?

Reply:
Yes, the lack of a logical operator of that form.
It could short circuit as soon as a true result is found.

Reply:
Bitwise OR won't short-circuit like Boolean OR does.  I thought maybe there's a chance that Microsoft inserted some weird special case logic when both operands are bool.  But I looked it up and they didn't.  Logical OR is guaranteed to short circuit but bitwise OR runs both operands, even for bool.  That's one of the reasons I wish there *was* a ``||=`` operator, so it could short circuit properly, but alas there isn't.